onnx/defs/math/defs.cc

```diff
@@ -955,29 +954,50 @@ ONNX_OPERATOR_SET_SCHEMA(
             axis += rank;
           if (axis < 0 || axis >= rank)
             fail_shape_inference("Invalid value for attribute axis");
-          // TODO: unclear what results should be if axis has less than k
-          // elements.
-          // Infer output shape if 'K' is available
+
+          const auto& axis_dim = input_shape.dim(static_cast<int>(axis));
           const auto* k = ctx.getInputData(1);
-          if (nullptr != k) {
-            if (k->dims_size() != 1 || k->int64_data_size() != 1 ||
-                k->data_type() != TensorProto::INT64)
+
+          // Infer output shape if:
+          // (1) 'K' is available
+          // (2) axis_dim has dim value
+          // Othewise cannot reliably compute output shape as axis dim value is
+          // unknown and hence cannot determine if axis dim value >= k (which
+          // should be enforced)
+          if (nullptr != k && axis_dim.has_dim_value()) {
+            int64_t k_value = 0;
+            if (k->dims_size() != 1 || k->dims(0) != 1)
+              fail_shape_inference(
+                  "K input must be a one-dimensional tensor of size 1.");
+            if (k->has_raw_data() && k->data_type() == TensorProto::INT64)
```

Comment:
Thank you!

We may do a little bit refactoring to have a function (template) to get data from TensorProto. That function will cover the raw_data check, otherwise, all shape inference functions (using input data / initializers) will have these two "if"s.

Reply:
Thanks @linkerzhang. I refactored the function that was previously only parsing raw data to generically handle data parsing (not just raw). So, two "if"s should go away now.